src/poll.rs

```diff
@@ -360,29 +634,79 @@ pub fn selector(poll: &Poll) -> &sys::Selector {
 
 impl Registration {
     /// Create a new `Registration` associated with the given `Poll` instance.
+    ///
     /// The returned `Registration` will be associated with this `Poll` for its
-    /// entire lifetime.
-    pub fn new(poll: &Poll, token: Token, interest: Ready, opts: PollOpt) -> (Registration, SetReadiness) {
-        let inner = RegistrationInner::new(poll, token, interest, opts);
-        let registration = Registration { inner: inner.clone() };
-        let set_readiness = SetReadiness { inner: inner.clone() };
+    /// entire lifetime. Dropping the `Registration` will prevent any further
+    /// notifications to be polled.
+    pub fn new(poll: &Poll, token: Token, interest: Ready, opt: PollOpt) -> (Registration, SetReadiness) {
+        is_send::<Registration>();
+        is_sync::<Registration>();
+        is_send::<SetReadiness>();
+        is_sync::<SetReadiness>();
+
+        // Clone handle to the readiness queue, this bumps the ref count
+        let queue = poll.readiness_queue.clone();
+
+        // Allocate the registration node. The new node will have `ref_count`
+        // set to 3, and `num_registration` set to
+        // 1. The 3 ref_counts represent ownership by one SetReadiness, one
+        // Registration, and the Poll handle.
```

Comment:
Hm ok so reading some more, I wonder if this is a source of a memory leak? If I create a `Registration` and `SetReadiness` pair, never call `update`, and then drop them, I think the refcount will still be at one and it'll leak?

Out of curiosity, was `Arc<ReadinessNode>` not possible here?